README.md
# AUTO-libs 凹凸公用库

1. [公用样式和脚本引入](#公用样式和脚本引入)
   - [reset 样式](#reset-样式)
   - [flexible 布局脚本](#flexible-布局脚本)
   - [Input 扩展](#input-扩展)
2. [CDN](#CDN-方法)
3. [Report 错误上报](#Report-错误上报)
4. [Search 浏览器地址参数处理](#Search-浏览器地址参数处理)
5. [go 跳转](#go-跳转)
   - [取还车地址](#取还车地址)
   - [支付](#支付)
   - [身份认证](#身份认证)
   - [驾照认证](#驾照认证)
6. [http](#http)
   - [自定义拦截器](#自定义拦截器)
7. [AS 统计埋点](#as-统计埋点)
   - [页面编号 事件号和携带参数](#页面编号-事件号和携带参数)
8. [WX](#wx)
   - [自定义参数](#自定义参数)
   - [自定义按钮](#自定义按钮)
9. [token 相关](#token-相关)
   - [token](#token)
   - [登录](#登录)
   - [openId](#openid)
   - [unionId](#unionid)
   - [virtualNo](#virtualno)
   - [memNo](#memno)
10. [时间转换](#时间转换)
    - [offsetHours](#offsetHours)
    - [offsetDays](#offsetDays)
    - [stringToDate](#stringToDate)
11. [httpConsole](#httpConsole)
    - [自定义拦截器 管理后台](#自定义拦截器)
    - [自定义登陆跳转 管理后台](#自定义登陆跳转)

## 公用样式和脚本引入

### reset 样式

```js
import 'auto-libs/build/styles/reset';
```

### flexible 布局脚本

```js
import 'auto-libs/build/scripts/flexible';
```

### Input 扩展

```js
import 'auto-libs/build/scripts/inputEvents';
```

## CDN 方法

获取 cdn 的静态资源或图片资源，cdn 地址会自动根据开发和生产环境切换

另外当传递的值为空时，返回一张灰白色空图，保证了不会访问 cdn/undefined 这样的地址

```js
import { CDN } from 'auto-libs';

CDN.image('myimage.jpg'); // https://carphoto.atzuche.com/myimage.jpg

CDN.asset('myimage.jpg'); // https://cdn.atzuche.com/myimage.jpg

CDN.asset(void 0); // https://cdn.atzuche.com/static/images/space.png
```

## Report 错误上报

上报的错误将会到 sentry 仅适用于 react-scrupts-auto 脚手架项目

```js
import { Report } from 'auto-libs';

// 用于做一些记录
Report.info('info消息');

// 用于警告
Report.warning('warning消息');

// 用于记录错误
Report.error(new Error('错误'));
```

1. 使用`Report.error`时注意：传递的内容不管是不是一个 Error 类型的数据，用`new Error()`包一层，方便 sentry 做错误定位
1. 一般我们在`try { ... } catch (err) { ... }`的 catch 里上报错误，全局的错误在该脚手架中会自动上报

## Search 浏览器地址参数处理

```js
import { Search } from 'auto-libs'

// 获取当前网址search内容的map
Search.parse<T>()

// 获取当前网址search内容的string
Search.toString()

// map格式转string格式
const str = Search.map2string({ a: 1, b: 2 }) // a=1&b=2

// string格式转map格式
const map = Search.string2map('a=1&b=2') // { a: '1', b: '2' }

// 获取当前网址search的某个key
const t = Search.get('token')

// 获取当前网址search的某个key，当值为undefined时，取默认值
const t = Search.getDefault('token', '112233')

// 设置某个值到当前网址search中
Search.set('token', '112233')

// 当该key不存在时，设置某个值到当前网址search中
// 注意：当key存在值，不做任何处理
Search.define('token', '112233')

// 删除当前网址search中的某个key
Search.remove('token')

// 清空当前网址的search内容
Search.clear()

// 判断某个key是否存在于当前search
const hasToken = Search.exist('token')

// 比较当前search中某个值
const isNew = Search.is('new', '1')
```

## go 跳转

```js
import { go } from 'auto-libs';
```

### 取还车地址

```js
const { address } = go

address({
  redirectUrl: string
  redirectParam: string
  locationInfo?: {
    address?: string;
    name?: string;
    code?: string;
    lat?: string;
    ng?: string;
    telPrefix?: string;
  }
  telPrefix?: string
  cityCode?: string
  keyword?: string
  servicePoint?: 1 | 0
  searchTip?: string
  selectCity?: 1 | 0
  selectCityTip?: string
  isOnDoor?: 1 | 0
})
```

### 支付

```js
const { pay } = go

pay({
  amount?: string
  payTitle?: string
  token: string
  orderNo: string
  payKind: string
  redirect_url: string
})

```

### 身份认证

```js
const { identityAuth } = go

identityAuth({
  redirect?: string
})
```

### 驾照认证

```js
const { dirverAuth } = go

dirverAuth({
  redirect?: string
})
```

## http

```js
import { http } from 'auto-libs';

http.request({
  method: 'get',
  url: `/wechatgw/jsapi/ticket`,
});
```

## httpConsole

```js
import { httpConsole } from 'auto-libs';

httpConsole.request({
  method: 'get',
  url: `/wechatgw/jsapi/ticket`,
});
```

### 自定义拦截器

```js
import { http } from 'auto-libs';

http.interceptors.request.use(config => {
  if (config.method === 'get') {
    if (typeof config.params !== 'object') {
      config.params = {};
    }

    config.params.requestId = Number(new Date());
  }

  if (Object.keys(config.data).length > 0) {
    config.data.requestId = Number(new Date());
  }

  return config;
});

export default http;
```

### 自定义登陆跳转

```js
import { httpConsole, httpConsoleExtend } from 'auto-libs';

httpConsoleExtend({
  toLogin: config => {
    if (config.status === 400) {
      location.href = '/login';
      return true;
    }

    return false;
  },
});
```

## AS 统计埋点

```js
import { AS } from 'auto-libs';

AS();
```

### 页面编号 事件号和携带参数

```js
AS({
  pageNo: 100,
  eventNo: 1000,
  eventContent: {
    name: 'hello world',
  },
});
```

## WX

```js
import { WX } from 'auto-libs';

// 微信分享
WX.share();
```

### 自定义参数

```js
WX.share({
  shareTitle: '凹凸租车', // 分享标题
  url: 'https://m.aotuzuche.com', // 分享链接，该链接域名或路径必须与当前页面对应的公众号JS安全域名一致
  sharePicUrl: 'https://carphoto.aotuzuche.com/web/auto/assets/imgs/logo.png', // 分享图标
  text: '凹凸租车 共享' // 分享详情
  type?: "userinfo" | "base" // 弹窗授权或者静默授权，默认静默
})
```

### 自定义按钮

```js
WX.share({
  shareTitle: '凹凸租车', // 分享标题
  url: 'https://m.aotuzuche.com', // 分享链接，该链接域名或路径必须与当前页面对应的公众号JS安全域名一致
  sharePicUrl: 'https://carphoto.aotuzuche.com/web/auto/assets/imgs/logo.png', // 分享图标
  text: '凹凸租车 共享', // 分享详情
  type?: "userinfo" | "base", // 弹窗授权或者静默授权，默认静默
  jsApiList?: jsApiListType[],  // menus
  callback?: (wx) => void  // ready 回调
})
```

## token 相关

### token

```js
import {
  getToken,
  setToken,
  clearToken,
  initToken,
  getConsoleToken,
  setConsoleToken,
  clearConsoleToken,
} from 'auto-libs';

// 获取 token
getToken();

// 设置 token
setToken();

// 清空 token
clearToken();

// 获取console token
getConsoleToken();

// 设置console token
setConsoleToken();

// 清空console token
clearConsoleToken();

// 初始化 token
initToken();

// 或者
await initToken();

// 忽略部分页面做initToken
initToken(() => {
  if (window.location.href === '需要忽略的url') {
    return true;
  }
  return false;
});
```

### 登录

```js
import { toLogin } from 'auto-libs'

toLogin(params?)
```

### openId

```js
import { getOpenId, setOpenId, clearOpenId } from 'auto-libs';

// 获取 openId
getOpenId();

// 设置 openId
setOpenId();

// 清空 openId
clearOpenId();
```

### unionId

```js
import { getUnionId, setUnionId, clearUnionId } from 'auto-libs';

// 获取 unionId
getUnionId();

// 设置 unionId
setUnionId();

// 清空 unionId
clearUnionId();
```

### virtualNo

```js
import { getVirtualNo, setVirtualNo, clearVirtualNo } from 'auto-libs';

// 获取虚拟会员号
getVirtualNo();

// 设置虚拟会员号
setVirtualNo();

// 清空虚拟会员号
clearVirtualNo();
```

### memNo

```js
import { getMemNo, setMemNo, clearMemNo } from 'auto-libs';

// 获取 memNo
getMemNo();

// 设置 memNo
setMemNo();

// 清空 memNo
clearMemNo();
```

## 时间转换

### offsetHours

### offsetDays

### stringToDate

```js
import { offsetHours, offsetDays, stringToDate } from 'auto-libs';

// 时间差，单位为小时
offsetHours();

// 将时间差转为文字方式
offsetDays();

// 将字符串20190101093000 转换为时间格式
stringToDate();
```

## 正则

添加 Reg 对象，包含下面这些方法和属性：

是否是凹凸 app: isApp,

是否是微信: isWX,

是否是支付宝客户端: isAlipay,

是否是 iOS: isiOS,

是否是 android: isAndroid,

是否是百度客户端： isBaidu,

是否是企业微信：isWXWork,

是否是手机号 function： isMobile(str),

是否是电话号码 function： isTel(str),

是否是电子邮箱 function： isEmail(str),

是否是身份证 function：isIDCard(str),

是否是支付宝小程序 function isMiniAlipay() 异步函数,

是否是微信小程序 function isMiniWX() 异步函数,

是否是微信小程序 function isMiniBaidu() 异步函数,


CHANGELOG.md
## [0.3.37](https://github.com/aotuzuche/auto-libs/compare/v0.3.30...v0.3.37) (2021-04-22)

### Bug Fixes

- 修改小程序 openWebview ([12b8e63](https://github.com/aotuzuche/auto-libs/commit/12b8e63e99128b1851394d7b3608384e7f6a10da))
- 修改小程序 openWebview ([eb87dd7](https://github.com/aotuzuche/auto-libs/commit/eb87dd7afbf52c03fbef394dba0ac5c3700108a9))

### Features

- http 增加 onLogin 方法 ([e4efa3a](https://github.com/aotuzuche/auto-libs/commit/e4efa3a44d441b2d91b9e3d434e450ea40e32e64))
- update version ([970f9c6](https://github.com/aotuzuche/auto-libs/commit/970f9c6d07bc97bfc58a0091d1a65d61c02c19ec))
- 修改 axios 版本 ([fcfbc9d](https://github.com/aotuzuche/auto-libs/commit/fcfbc9d040ef234897d753ab90d5ef158a5c5b21))
- 修改 axios 版本 ([6d0b9c1](https://github.com/aotuzuche/auto-libs/commit/6d0b9c1bb7ed285bc3b33b0ff780ff162d6e8987))
- 修改 http 方法 ([4c1b2ce](https://github.com/aotuzuche/auto-libs/commit/4c1b2ceaf71c1d734e3f5574438a08fc0a7b3ad7))
- 修改类型 ([bb34a56](https://github.com/aotuzuche/auto-libs/commit/bb34a56b1f3056dfecea08bcb039182e3d548993))
- 添加 types ([0be720e](https://github.com/aotuzuche/auto-libs/commit/0be720ee2bc4bf6dede64e95285ca38997fd8140))

## [0.3.30](https://github.com/aotuzuche/auto-libs/compare/v0.3.29...v0.3.30) (2021-03-18)

### Bug Fixes

- 🐛 解决 eslint 问题 ([820f696](https://github.com/aotuzuche/auto-libs/commit/820f696959926300cf5c5d9fd3914446b985a754))

### Features

- 🎸 添加隐藏手机号功能 ([934cfed](https://github.com/aotuzuche/auto-libs/commit/934cfed4e62ff272d4115341236ef10621b1f2cd))
- as 方法增加 uuid ([04965b4](https://github.com/aotuzuche/auto-libs/commit/04965b443a8de33b07ff67572980facd7b11cf98))
- go.pay 的 redirect_url 改为非必填 ([19a013b](https://github.com/aotuzuche/auto-libs/commit/19a013b4965651665ecf6d74bccaa8390ebb3935))
- 增加 httpConsoleExtend ([b71f98f](https://github.com/aotuzuche/auto-libs/commit/b71f98f1291dd6216bbd72806a4bffe56bc5b39b))
- 增加小程序登录 ([faa06e4](https://github.com/aotuzuche/auto-libs/commit/faa06e4cecb402e69f9c756425c7dce5f3b34670))
- 小程序 naviteTo 方法当跳转 tab 页面时,使用 switchTab ([5b7ba8f](https://github.com/aotuzuche/auto-libs/commit/5b7ba8fbb4ad9937f5d75a635fd518116adbf21f))
- 小程序 naviteTo 方法当跳转 tab 页面时,使用 switchTab ([dd736b3](https://github.com/aotuzuche/auto-libs/commit/dd736b3cbbff9e66d5f5ccd91adbedc71fb7eb52))
- 小程序内的 h5 在 toLogin 时调用原生小程序登录 ([f73c3d6](https://github.com/aotuzuche/auto-libs/commit/f73c3d650f6ccbdd2714487c5ea557bb2437ce72))

## [0.3.22](https://github.com/aotuzuche/auto-libs/compare/v0.3.15...v0.3.22) (2020-12-10)

### Bug Fixes

- tslint fail ([4abb630](https://github.com/aotuzuche/auto-libs/commit/4abb630e6b512e2e654d3e57986bf10153e413e7))

### Features

- at-js-sdk 改用 at-js-bridge ([5573ac5](https://github.com/aotuzuche/auto-libs/commit/5573ac518497238b18f89170ee9c214de2c36efb))
- mini 增加 jd 小程序的支持,增加 postMessage 方法的封装 ([5d099e8](https://github.com/aotuzuche/auto-libs/commit/5d099e8575bad54e493772d3e2b491085695e9fc))
- mini 增加 jd 小程序的支持,增加 postMessage 方法的封装 ([2c9793e](https://github.com/aotuzuche/auto-libs/commit/2c9793e3648578a9cfb3b2fd60d61128707b27be))
- Mini 增加 switchTab 方法 ([364cfee](https://github.com/aotuzuche/auto-libs/commit/364cfee3203d8073920d93a0c3b4d7881abf7a51))
- 增加 httpConsole report ([194a943](https://github.com/aotuzuche/auto-libs/commit/194a943c8bc43491570f6a1945ee2eaa09bbe45e))
- 放开 report 方法的使用限制 ([ca6fbe6](https://github.com/aotuzuche/auto-libs/commit/ca6fbe6b806f3aac36cf273dd83451d8dd8a7858))

## [0.3.15](https://github.com/aotuzuche/auto-libs/compare/v0.3.4-beta.0...v0.3.15) (2020-12-02)

### Bug Fixes

- httpConsole 401 status toLogin ([9377cd2](https://github.com/aotuzuche/auto-libs/commit/9377cd2694a939d2abc498985985c1ae33b76fd5))
- 暂时删除 http 中的 report 功能 ([21fac1d](https://github.com/aotuzuche/auto-libs/commit/21fac1df9c7c61d0f3dfa60054bdae0de11183aa))

### Features

- as 方法增加小程序支持 ([aaa5570](https://github.com/aotuzuche/auto-libs/commit/aaa55708c1172e948e1068a11edb6413ab15d3df))
- as 方法增加小程序支持 ([ce9cc71](https://github.com/aotuzuche/auto-libs/commit/ce9cc71428b36db64a2bbd287d73ff44324d77cc))
- error report ([ee5d653](https://github.com/aotuzuche/auto-libs/commit/ee5d653256f5b023bb78e8cdd9e9cf010edc6010))
- 修改 report 上报的方式 ([f4c0b64](https://github.com/aotuzuche/auto-libs/commit/f4c0b64e8d0ba49709e7ac3c858f010f205cf690))
- 发布 beta 版本 ([4b06e1d](https://github.com/aotuzuche/auto-libs/commit/4b06e1d3ce10a150e6244b90b47c29b9fa9921ef))
- 处理 search 方法的类型, <T extends any>改为<T = any>, report 上报错误时忽略空消息 ([a191ffb](https://github.com/aotuzuche/auto-libs/commit/a191ffbd218a5fd2b5a013924e056162c2abe9c8))
- 对京东小程序增加 Reg.isJD ([410eb9d](https://github.com/aotuzuche/auto-libs/commit/410eb9db83cdd51531d00f945e61eef36392d1ae))
- 当 window 下有 platform 时,http 携带上 ([ab3cfac](https://github.com/aotuzuche/auto-libs/commit/ab3cfac3102ddddfecb41903189de0fbe87078aa))
- 支付增加 amount 和 payTitle 字段 ([6b19fa5](https://github.com/aotuzuche/auto-libs/commit/6b19fa5b810e95e4815608d2f5efc212ff5c6f14))
- 添加 api 请求上报功能 ([4a2bd93](https://github.com/aotuzuche/auto-libs/commit/4a2bd93ad9ba0aa15ee221813ab8a950e282f2f1))
- 统计数据增加 utm ([1a9f670](https://github.com/aotuzuche/auto-libs/commit/1a9f6700ef2c11329041d52abcecc77488651fc3))
- 统计数据增加 utm ([c309f02](https://github.com/aotuzuche/auto-libs/commit/c309f02f8e3b1d6fe6c2a9d2db07f079eb8ac9de))
- 统计数据增加 utm ([597ba16](https://github.com/aotuzuche/auto-libs/commit/597ba16b26957612decffac9fa661367cf61f84a))
- **mini:** 增加小程序相关的方法封装 ([200ab03](https://github.com/aotuzuche/auto-libs/commit/200ab030a73e2f2d2cd7fa76761295b508dc004c))
- **mini:** 增加小程序相关的方法封装 ([b80deb2](https://github.com/aotuzuche/auto-libs/commit/b80deb2d1cf0cdd289fc17865989e6dad7d908ca))
- **mini:** 增加小程序相关的方法封装 ([16c93a3](https://github.com/aotuzuche/auto-libs/commit/16c93a3e22977bec96801f0fe646c89d06d2b136))
- **mini:** 增加小程序相关的方法封装 ([66a7bda](https://github.com/aotuzuche/auto-libs/commit/66a7bda7c6f353afb0e51314cd0a4c0bcdb07acb))

## [0.3.4-beta.0](https://github.com/aotuzuche/auto-libs/compare/v0.3.3...v0.3.4-beta.0) (2020-07-13)


lib/http.ts
import axios, { AxiosAdapter, AxiosInstance, AxiosPromise, AxiosRequestConfig } from 'axios';
import { clearToken, getToken, toLogin } from './token';
import report from './utils/analyticsReport';
import getUtm from './utils/getUtm';

interface HttpConfig {
  resCode?: string;
  resMsg?: string;
  data?: any;
}

export interface ProcessEnv {
  [key: string]: string | undefined;
}

class HttpError extends Error {
  public msg: string;
  public name = 'HttpError';
  public data: any;
  public code = '0';
  public constructor(message: string, data?: HttpConfig) {
    super(message);

    this.msg = message;
    if (data) {
      this.data = data ? (data.data ? data.data : data) : null;
      this.code = data.resCode || '';
    }
  }

  public toString() {
    return this.message;
  }
}

// 用户唯一id
let uuid = localStorage.getItem('_app_uuid_');
if (!uuid) {
  uuid = String(new Date().valueOf()) + Math.round(Math.random() * 9999);
  localStorage.setItem('_app_uuid_', uuid);
}

/**
 * 配置axios
 */

interface CustomRequestConfig extends AxiosRequestConfig {
  onLogin?: (code: string) => void;
  report?: boolean;
}

interface CustomAxiosInstance extends AxiosInstance {
  request<T = any>(config: CustomRequestConfig): AxiosPromise<T>;
}

export const http: CustomAxiosInstance = axios.create({
  baseURL: '/',
  headers: {
    Accept: 'application/json;version=3.0;compress=false',
    'Content-Type': 'application/json;charset=utf-8',
  },
  data: {},
});

/**
 * 请求拦截器，在发起请求之前
 */
http.interceptors.request.use(config => {
  const token = getToken();
  const platform = (window as any).platform;
  const utm = getUtm();

  const method = (config.method as string).toLocaleLowerCase();
  if (token) {
    config.headers['Atzuche-Token'] = token;
  }
  if (method === 'get') {
    if (typeof config.params !== 'object') {
      config.params = {};
    }

    // 兼容appserver的接口，appserver的接口token需要带在参数中，post请求也是一样
    if (token) {
      config.params.token = token;
    }

    if (utm) {
      config.params.utmSource = utm.utm_source;
      config.params.utmMedium = utm.utm_medium;
      config.params.utmCampaign = utm.utm_campaign;
      config.params.utmTerm = utm.utm_term;
    }

    config.params.requestId = Number(new Date());
    if (platform) {
      config.params.h5Platform = platform;
    }
  }

  const methods: string[] = ['post', 'put', 'patch', 'delete'];
  if (methods.indexOf(method) > -1 && typeof config.data !== 'string') {
    if (token) {
      config.data.token = token;
    }

    if (utm) {
      config.data.utmSource = utm.utm_source;
      config.data.utmMedium = utm.utm_medium;
      config.data.utmCampaign = utm.utm_campaign;
      config.data.utmTerm = utm.utm_term;
    }

    config.data.requestId = Number(new Date());
    if (platform) {
      config.data.h5Platform = platform;
    }
  }

  (config as any).____t = new Date().valueOf();

  return config;
});

/**
 * 接口响应拦截器，在接口响应之后
 */
http.interceptors.response.use(
  config => {
    // config data
    const cc = (config.config as any) || {};

    // report
    report(cc.report !== false && cc.____t && cc.url && cc.method, 'api/m', {
      m: cc.method,
      u: cc.url,
      p: window.location.origin + window.location.pathname,
      s: (new Date().valueOf() - cc.____t) / 1000,
      uu: uuid,
    });

    let strictModel = true; // 严格模式
    const data = config.data || {};

    // 目前的判断方式：因为resCode与resMsg是java端必给的字段，所以认为没有该两个字段时，走标准的http status模式
    if (typeof data.resCode !== 'undefined' && typeof data.resMsg !== 'undefined') {
      strictModel = false;
    }

    if (strictModel) {
      if (config.status >= 200 && config.status < 300) {
        return data;
      }

      if (config.status === 401) {
        if (cc.onLogin) {
          cc.onLogin('200008');
        } else {
          clearToken();
          toLogin();
        }
        return {};
      }

      // report error
      report(cc.report !== false && cc.url && cc.method, 'api_error/m', {
        m: cc.method,
        u: cc.url,
        p: window.location.origin + window.location.pathname,
        uu: uuid,
        s: config.status,
        err: JSON.stringify(data),
      });

      return Promise.reject(new HttpError(data.message || '', data));
    }

    // atcz java端的模式

    // 响应正常
    if (data.resCode === '000000') {
      return data.data;
    }

    if (data.resCode === '200008') {
      // 需要登录（没登录或登录过期）
      if (cc.onLogin) {
        cc.onLogin(data.resCode);
      } else {
        clearToken();
        toLogin();
      }
      return {};
    }

    if (data.resCode === '200101') {
      // 需要绑定
      if (cc.onLogin) {
        cc.onLogin(data.resCode);
      } else {
        clearToken();
        toLogin({ isBind: true });
      }
      return {};
    }

    // report error
    report(cc.report !== false && cc.url && cc.method && data.resCode === '999999', 'api_error/m', {
      m: cc.method,
      u: cc.url,
      p: window.location.origin + window.location.pathname,
      uu: uuid,
      s: config.status,
      err: data.resMsg || data.msg || data.message,
    });

    // reject错误处理
    return Promise.reject(new HttpError(data.resMsg || data.msg || data.message, data));
  },
  error => {
    console.error('http:reject', error);

    if (error.response.status === 401) {
      clearToken();
      toLogin();
      return false;
    }

    const cc = error.config;
    const res = error.response || {};
    report(cc && cc.report !== false && res.status, 'api_error/m', {
      m: cc.method,
      u: cc.url,
      p: window.location.origin + window.location.pathname,
      uu: uuid,
      s: res.status,
      err: JSON.stringify(res.data),
    });

    // reject错误处理
    return Promise.reject(new HttpError(res.data && (res.data.message || '系统错误')));
  },
);

const httpCacheAdapter = (page: string, key: string, hour = 0) => {
  const adapter: AxiosAdapter = (conf: any) => {
    // 判断是否存在缓存数据
    const fullKey = `_auto_cache_${page}_${key}_`;
    let cache: string | null | undefined = void 0;
    let cacheJson: Record<string, any> = {};
    if (hour <= 0) {
      if (!(window as any)._auto_cache_) {
        (window as any)._auto_cache_ = {};
      } else {
        cache = (window as any)._auto_cache_[fullKey] ? '_' : void 0;
        cacheJson = (window as any)._auto_cache_[fullKey] || {};
      }
    } else {
      cache = localStorage.getItem(fullKey);
      cacheJson = cache ? JSON.parse(cache) : {};
      if (cacheJson.created && cacheJson.duration) {
        const horOffset = (new Date().valueOf() - cacheJson.created) / 1000 / 60 / 60;
        if (horOffset > cacheJson.duration) {
          localStorage.removeItem(fullKey);
          cache = null;
          cacheJson = {};
        }
      } else {
        cache = null;
      }
    }

    // 调用默认请求接口, 发送正常请求及返回
    if (!cache) {
      const newConf = { ...conf };
      delete newConf.adapter;
      return new Promise((resolve, reject) => {
        axios(newConf)
          .then(json => {
            const data = json.data || {};
            if (
              json.status === 200 &&
              data.resCode &&
              data.resMsg &&
              data.data &&
              data.resCode === '000000'
            ) {
              if (hour <= 0) {
                if (!(window as any)._auto_cache_) {
                  (window as any)._auto_cache_ = {};
                }
                (window as any)._auto_cache_[fullKey] = {
                  created: new Date().valueOf(),
                  response: data,
                  status: json.status,
                };
              } else {
                localStorage.setItem(
                  fullKey,
                  JSON.stringify({
                    created: new Date().valueOf(),
                    duration: hour,
                    response: data,
                    status: json.status,
                  }),
                );
              }
            }
            resolve(json);
          })
          .catch(err => reject(err));
      });
    }

    // 返回缓存数据
    return new Promise(resolve => {
      resolve({ data: cacheJson.response, status: cacheJson.status } as any);
    });
  };
  return adapter;
};

export { httpCacheAdapter };
